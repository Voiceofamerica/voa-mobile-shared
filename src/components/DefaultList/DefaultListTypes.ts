
import { ListItem } from '../../helpers/itemList'

export { ListItem }

export interface BaseProps {
  items: ListItem[] | undefined
  onItemClick: (id: number) => void
  className?: string
  style?: React.CSSProperties
}
