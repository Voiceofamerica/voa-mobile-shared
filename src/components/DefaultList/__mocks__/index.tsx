
import * as React from 'react'

function DefaultList (props) {
  return (
    React.createElement('default-list', { props }, props.children)
  )
}

namespace DefaultList {
  export const Static = (props) => (
    React.createElement('static-default-list', { props }, props.children)
  )
}

export default DefaultList
