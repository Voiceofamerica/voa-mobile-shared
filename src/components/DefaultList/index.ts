
export { default } from './DefaultList'
export * from './DefaultList'
export * from './DefaultListTypes'
