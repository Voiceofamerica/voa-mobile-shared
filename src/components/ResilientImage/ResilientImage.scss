
@import '../../commonStyles/index.scss';

.resilientImage {
  position: relative;
  background: black;

  .imageContent {
    object-fit: cover;
    width: 100%;
    height: 100%;

    &.containImage {
      object-fit: contain;
      height: auto;
      max-height: 100%;
    }
  }

  .childrenArea {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .spinner {
    color: $voa-blue;
    height: 80%;
    max-height: 100px;
    margin-top: -1%;
  }
}
