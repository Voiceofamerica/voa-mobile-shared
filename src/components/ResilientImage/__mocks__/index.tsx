
import * as React from 'react'

export default jest.fn(({ children, ...props}) => (<div {...props}>{children}</div>))
