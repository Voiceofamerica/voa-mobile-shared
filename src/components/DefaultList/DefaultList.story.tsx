
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import StaticDefaultList from './StaticDefaultList'
import DefaultList from './DefaultList'
import { ListItem } from './DefaultListTypes'

const testItem: ListItem = {
  id: 1,
  image: {
    tiny: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png',
    thumb: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png',
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const getIcon = (idx: number) => {
  idx = idx % 4
  if (idx === 0) {
    return 'video'
  } else if (idx === 1) {
    return 'audio'
  } else if (idx === 2) {
    return 'photoGallery'
  } else {
    return undefined
  }
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
      icon: getIcon(i),
    })
  }
  return output
}

const list = createList(1000)
const shortList = createList(30)

const containerStyle: React.CSSProperties = {
  height: '100vh',
  display: 'flex',
  flexDirection: 'column',
}

storiesOf(DefaultList.name, module)
  .add('display', () => (
    <div style={containerStyle}>
      <style>{'body { margin: 0; }'}</style>
      <DefaultList
        items={list}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <DefaultList
        items={list}
        onItemClick={action('onItemClick')}
        dir='rtl'
      />
    </div>
  ))
  .add('static height display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <DefaultList
        items={list}
        onItemClick={action('onItemClick')}
        height={200}
      />
    </div>
  ))
  .add('contained height display', () => (
    <div style={{height: 300, display: 'flex'}}>
      <style>{'body { margin: 0; }'}</style>
      <DefaultList
        items={list}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))

storiesOf(`${DefaultList.name}.Static`, module)
  .add('display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <StaticDefaultList
        items={shortList}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div dir='rtl'>
      <style>{'body { margin: 0; }'}</style>
      <StaticDefaultList
        items={shortList}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
