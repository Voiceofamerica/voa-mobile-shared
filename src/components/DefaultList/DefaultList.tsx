
import * as React from 'react'
import { List, ListRowProps, Index } from 'react-virtualized'
import { Subscription } from 'rxjs/Subscription'

import { resizeObservable } from '../../helpers/window'
import { TextDirection, directionObservable } from '../../helpers/text'
import Card, { getHeight as getCardHeight } from '../Card'
import SecondaryCard, { getHeight as getSecondaryCardHeight } from '../SecondaryCard'
import Ticket, { TICKET_HEIGHT } from '../Ticket'

import StaticDefaultList from './StaticDefaultList'
import { BaseProps } from './DefaultListTypes'
import { defaultList, row, ticketRow } from './DefaultList.scss'

export interface Props extends BaseProps {
  emptyContent?: string | JSX.Element
  height?: number
  width?: number
  dir?: TextDirection
}

export interface State {
  renderHeight: number
  renderWidth: number
  cardHeight: number
  secondaryCardHeight: number
  dir: TextDirection
}

const MENU_HEIGHT = 150
const BORDER_SIZE = 1
// 4px border-bottom + 7px margin-bottom
const CARD_MARGIN = 11

class DefaultList extends React.PureComponent<Props, State> {
  static Static = StaticDefaultList

  state: State = {
    renderHeight: window.innerHeight - MENU_HEIGHT,
    renderWidth: this.props.height || window.innerWidth,
    cardHeight: getCardHeight(),
    secondaryCardHeight: getSecondaryCardHeight(),
    dir: this.props.dir || directionObservable.getValue(),
  }

  private container: HTMLDivElement | null = null
  private virtualList: List | null = null
  private resizeSub: Subscription
  private directionSub: Subscription

  componentDidMount () {
    this.resizeSub = resizeObservable.subscribe(() => {
      const renderHeight = this.getNewHeight(this.props)
      const renderWidth = this.getNewWidth(this.props)
      const cardHeight = getCardHeight(renderWidth)
      const secondaryCardHeight = getSecondaryCardHeight(renderWidth)
      this.setState({ renderHeight, renderWidth, cardHeight, secondaryCardHeight }, () => {
        if (this.virtualList) {
          this.virtualList.recomputeRowHeights()
        }
      })
    })
    this.directionSub = directionObservable.subscribe(dir => {
      if (this.props.dir) {
        return
      }

      this.setState({ dir })
    })
  }

  componentWillUnmount () {
    this.resizeSub.unsubscribe()
    this.directionSub.unsubscribe()
  }

  componentWillReceiveProps (nextProps: Props) {
    const renderHeight = this.getNewHeight(nextProps)
    const renderWidth = this.getNewWidth(nextProps)
    const cardHeight = getCardHeight(renderWidth)
    const secondaryCardHeight = getSecondaryCardHeight(renderWidth)
    const dir = nextProps.dir || directionObservable.getValue()
    this.setState({ renderHeight, renderWidth, cardHeight, secondaryCardHeight, dir })
  }

  render () {
    const { items = [], emptyContent = '', className = '', style } = this.props

    return (
      <div ref={this.setContainer} className={`${defaultList} ${className}`} style={style}>
        {
          items.length > 0
          ? this.renderVirtualContent()
          : emptyContent
        }
      </div>
    )
  }

  private getNewHeight = ({ height = 0 }: Props) => {
    if (height > 0) {
      return height
    } else if (this.container) {
      return this.container.clientHeight
    } else {
      return window.innerHeight - MENU_HEIGHT
    }
  }

  private getNewWidth = ({ width = 0 }: Props) => {
    if (width > 0) {
      return width
    } else if (this.container) {
      return this.container.clientWidth
    } else {
      return window.innerWidth
    }
  }

  private renderVirtualContent = () => {
    const { items = [] } = this.props
    const { renderHeight, renderWidth } = this.state

    const count = items.length <= 2 ? items.length : items.length - 1

    return (
      <List
        ref={this.setVirtualList}
        height={renderHeight}
        rowHeight={this.getHeight}
        rowCount={count}
        width={renderWidth}
        rowRenderer={this.renderRow}
      />
    )
  }

  private getHeight = ({ index }: Index) => {
    if (index === 0) {
      return this.state.cardHeight + CARD_MARGIN
    } else if (index === 1) {
      return this.state.secondaryCardHeight
    } else {
      return TICKET_HEIGHT + BORDER_SIZE
    }
  }

  private renderRow = (rowProps: ListRowProps) => {
    if (rowProps.index === 0) {
      return this.renderCardRow(rowProps)
    } else if (rowProps.index === 1) {
      return this.renderSecondaryCardRow(rowProps)
    } else {
      return this.renderTicketRow(rowProps)
    }
  }

  private renderCardRow = ({ index, key, style }: ListRowProps) => {
    const { dir } = this.state
    const { items, onItemClick } = this.props

    const { id, image, title, icon } = items![0]

    return (
      <div key={key} className={row} style={style} dir={dir}>
        <Card
          onPress={() => onItemClick(id)}
          title={title}
          imageUrl={image && image.hero}
          icon={icon}
        />
      </div>
    )
  }

  private renderSecondaryCardRow = ({ key, style }: ListRowProps) => {
    const { dir } = this.state

    return (
      <div key={key} className={row} style={style} dir={dir}>
        {this.renderSecondaryCard(1)}
        {this.renderSecondaryCard(2)}
      </div>
    )
  }

  private renderSecondaryCard = (index: number) => {
    const { items, onItemClick } = this.props

    const item = items![index]

    if (!item) {
      return <div style={{ flex: 1 }} />
    }

    const { id, image, title, icon } = item

    return (
      <SecondaryCard
        onPress={() => onItemClick(id)}
        title={title}
        imageUrl={image && image.thumb}
        icon={icon}
      />
    )
  }

  private renderTicketRow = ({ index, isScrolling, key, style }: ListRowProps) => {
    const { dir } = this.state
    const { items, onItemClick } = this.props

    const { id, image, title, minorText, icon } = items![index + 1]

    return (
      <div key={key} className={ticketRow} style={style} dir={dir}>
        <Ticket
          onPress={() => onItemClick(id)}
          title={title}
          imageUrl={image && image.tiny}
          minorText={minorText}
          icon={icon}
          suppressImage={isScrolling}
        />
      </div>
    )
  }

  private setContainer = (el: HTMLDivElement | null) => {
    this.container = el
    const renderHeight = this.getNewHeight(this.props)
    const renderWidth = this.getNewWidth(this.props)
    const cardHeight = getCardHeight(renderWidth)
    const secondaryCardHeight = getSecondaryCardHeight(renderWidth)
    this.setState({ renderHeight, renderWidth, cardHeight, secondaryCardHeight })
  }

  private setVirtualList = (el: List | null) => {
    this.virtualList = el
  }
}

export default DefaultList
