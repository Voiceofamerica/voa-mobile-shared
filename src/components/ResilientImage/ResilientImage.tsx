
import * as React from 'react'

import { waitUntilOnline } from '../../helpers/resilience'

import Spinner from '../Spinner'

import { resilientImage, imageContent, childrenArea, containImage, spinner } from './ResilientImage.scss'

export interface Props {
  src: string | null
  showSpinner?: boolean
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
  onClick?: () => void
  contain?: boolean
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imgUrl?: string
  imageStaus: ImageStatus
}

const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  mounted: boolean = false

  readonly abortController = new AbortController()

  componentWillMount () {
    this.mounted = true
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  componentWillUnmount () {
    this.mounted = false
    this.abortController.abort()
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.src !== this.props.src) {
      this.setImageStatus('loading')

      const shouldRender = navigator.onLine || this.props.alwaysShow
      this.setState({
        shouldRender,
      })

      if (shouldRender) {
        this.tryFetchImage()
          .catch(err => {
            console.warn(`Something went wrong loading ${nextProps.src}`, err)
          })
      }
    }
  }

  tryFetchImage = async () => {
    const { src } = this.props

    if (!src || !this.mounted) {
      return
    }

    if (!navigator.onLine) {
      waitUntilOnline().then(() => this.mounted && this.tryFetchImage())
      return
    }

    const { onLoadDone = noop } = this.props

    try {
      const blobUrl = await fetch(src, {
          method: 'GET',
          signal: this.abortController.signal,
        })
        .then(res => res.blob())
        .then(blob => URL.createObjectURL(blob))

      this.setImageStatus('loaded', blobUrl)
      onLoadDone()

    } catch (err) {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    }
  }

  setImageStatus = (imageStaus: ImageStatus, imgUrl?: string) => {
    if (!this.mounted) {
      return
    }

    this.setState({
      imageStaus,
      imgUrl,
    })
  }

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus, imgUrl } = this.state
    const { defaultSrc = require('./imagedefault.gif'), contain } = this.props

    const isLoaded = imageStaus === 'loaded'
    const srcToUse = isLoaded && imgUrl
                   ? imgUrl
                   : defaultSrc

    const className = contain ? `${imageContent} ${containImage}` : imageContent

    return (<img src={srcToUse} className={className} />)
  }

  renderSpinner = () => {
    const { showSpinner } = this.props
    const { imageStaus } = this.state

    const isLoaded = imageStaus === 'loaded'

    if (isLoaded || !showSpinner) {
      return null
    }

    return (
      <div className={childrenArea}>
        <Spinner className={spinner} />
      </div>
    )
  }

  renderChildren = () => {
    const { children } = this.props

    return (
      <div className={childrenArea}>
        {children}
      </div>
    )
  }

  render () {
    const { shouldRender = navigator.onLine || this.props.alwaysShow } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
        onClick = () => null,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style} onClick={ev => onClick()}>
          { this.renderContent() }
          { this.renderSpinner() }
          { this.props.children ? this.renderChildren() : null }
        </div>
      )
    }
  }
}

export default ReilientImage
