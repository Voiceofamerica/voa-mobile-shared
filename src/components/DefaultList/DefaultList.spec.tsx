
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { ListItem } from './DefaultListTypes'
import DefaultList from './DefaultList'
import ResilientImage from '../ResilientImage'

jest.mock('../ResilientImage')
const testItem: ListItem = {
  id: 1,
  image: {
    tiny: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png',
    thumb: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png',
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
    })
  }
  return output
}

const list = createList(10)

describe(`<${DefaultList.name} />`, () => {
  describe('shapshots', () => {
    it('should render all of the items', () => {
      const element = create((
        <DefaultList items={list} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render nothing if no items', () => {
      const element = create((
        <DefaultList items={[]} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should not throw if items is undefined', () => {
      const element = create((
        <DefaultList items={undefined} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
