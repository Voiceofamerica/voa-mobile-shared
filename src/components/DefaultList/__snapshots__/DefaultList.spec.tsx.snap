// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DefaultList /> shapshots should not throw if items is undefined 1`] = `
<div
  className="defaultList "
  style={undefined}
>
  
</div>
`;

exports[`<DefaultList /> shapshots should render all of the items 1`] = `
<div
  className="defaultList "
  style={undefined}
>
  <div
    aria-label="grid"
    aria-readonly={true}
    className="ReactVirtualized__Grid ReactVirtualized__List"
    id={undefined}
    onScroll={[Function]}
    role="grid"
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "boxSizing": "border-box",
        "direction": "ltr",
        "height": 618,
        "overflowX": "hidden",
        "overflowY": "auto",
        "position": "relative",
        "width": 1024,
        "willChange": "transform",
      }
    }
    tabIndex={0}
  >
    <div
      className="ReactVirtualized__Grid__innerScrollContainer"
      role="rowgroup"
      style={
        Object {
          "height": 1685.68,
          "maxHeight": 1685.68,
          "maxWidth": 1024,
          "overflow": "hidden",
          "pointerEvents": "",
          "position": "relative",
          "width": "auto",
        }
      }
    >
      <div
        className="row"
        dir="ltr"
        style={
          Object {
            "height": 587,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
      >
        <div
          className="card undefined"
          onClick={[Function]}
          style={undefined}
        >
          <div
            alwaysShow={true}
            className="imageContainer"
            showSpinner={true}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png"
          />
          <div
            className="gradient"
          >
            <div
              className="titleText"
            >
              <span>
                 
                President Trump: From North Korea to Iran, we confront dangerous governments
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="row"
        dir="ltr"
        style={
          Object {
            "height": 363.68,
            "left": 0,
            "position": "absolute",
            "top": 587,
            "width": "100%",
          }
        }
      >
        <div
          className="secondaryCard undefined"
          onClick={[Function]}
          style={undefined}
        >
          <div
            alwaysShow={true}
            className="imageContainer"
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png"
          />
          <div
            className="titleText"
          >
             
            President Trump: From North Korea to Iran, we confront dangerous governments
          </div>
        </div>
        <div
          className="secondaryCard undefined"
          onClick={[Function]}
          style={undefined}
        >
          <div
            alwaysShow={true}
            className="imageContainer"
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png"
          />
          <div
            className="titleText"
          >
             
            President Trump: From North Korea to Iran, we confront dangerous governments
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 950.6800000000001,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1055.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1160.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1265.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1370.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1475.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
      <div
        className="ticketRow"
        dir="ltr"
        style={
          Object {
            "height": 105,
            "left": 0,
            "position": "absolute",
            "top": 1580.68,
            "width": "100%",
          }
        }
      >
        <div
          className="ticket"
          onClick={[Function]}
          style={
            Object {
              "height": 80,
              "padding": 12,
            }
          }
        >
          <div
            alwaysShow={true}
            className="imageArea"
            onLoadDone={[Function]}
            src="https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png"
          />
          <div
            className="content"
          >
            <div
              className="textContent"
            >
              <div
                className="titleText"
              >
                President Trump: From North Korea to Iran, we confront dangerous governments
              </div>
            </div>
            <div
              className="minorText"
            >
              15:08
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DefaultList /> shapshots should render nothing if no items 1`] = `
<div
  className="defaultList "
  style={undefined}
>
  
</div>
`;
